### ISSUE TITLE
re-enable metric with dag size sum for pins

### ISSUE THREAD

**Issue by vasco-santos** [web3-storage/web3.storage#833]
Revert https://github.com/web3-storage/web3.storage/pull/824 to gather the metrics again.

## Problem

This computation has been a bottleneck before and per analysis in Heroku Metrics it had been taking consistently 20+ seconds for the operation, each time the metrics endpoint was queried. With it disabled, we have been consistently responding in less than 10 seconds to every request, which also means the metrics queries are not blocking the DB temporarily.

For getting the sum of the dag size for the Pins (Pinned state), we need to go join the Table of Pins (as the time of writing with over 52M rows, 26M rows with Pinned state), merge it with the Content table to get the dag size of each Pin and then go through all the table, in order to sum the Dag Sizes. This solution does not seem to scale as we move forward and the tables continue to grow.

Running with pgsql with timing enabled shows that even not using pgrest the execution time is getting out of control:

```sql
select sum(c.dag_size)
    from pin p
    join content c on c.cid = p.content_cid
    where p.status = ('Pinned')::pin_status_type;
       sum
-----------------
 167639145461473
(1 row)

Time: 51838.613 ms (00:51.839)
```

We already have indexes to optimize this query:

```sql
CREATE UNIQUE INDEX content_cid_with_size_idx ON content (cid) INCLUDE (dag_size);
CREATE INDEX IF NOT EXISTS pin_content_cid_idx ON pin (content_cid);
CREATE INDEX IF NOT EXISTS pin_status_idx ON pin (status);
CREATE INDEX IF NOT EXISTS pin_composite_pinned_at_idx ON pin (content_cid, updated_at) WHERE status = 'Pinned';
```

## Solution

We talked about using the replica DB for these READ queries, but we would need to setup a new pgrest dyno and wire it up, and still be limited to 30s execution time (need to double check if we can configure this execution time.

To potential solutions to keep this metrics are:

- have a cumulative sum being computed over time and we just update it based on timestamps using a CRON. This might be error prone as we have seen before (Fauna) when dag sizes are computed async and then we don't know if it was already tracked in the Pin metrics or not. 
- always compute it but perform some sort of pagination to perform the query in multiple operations and sum all the results

## Alternative

This metric currently represents ~3x the dag size sum of the content table. It will change if we change our Cluster setup. We will likely even change it, but in a way that we will also need to rethink how we track the Pins with AWS IPFS.

Taking into account the above, we can probably consider computing the sum of dag size for `Pinning`, `PinError` and `PinQueued` instead of `Pinned` (we can get their sum in less than 1s). We can use the guideline of `~3x` the dag size sum of the content table (we have this metric), and use the remaining ones to identify potential problems that happen. We also have metrics that count the elements in each Pin state.

This alternative solution is not ideal and will not be accurate if we move away from the current cluster size. But probably that is unlikely until we move to AWS IPFS, where we also need to rethink how to get this metric anyway. So, this is probably a good enough short term solution with little effort to achieve.

**Comment by dchoi27** [comment]
Thanks for writing this up! I'm leaning towards the alternative being a sufficient solution. From a metrics standpoint, I think the main reason we'd want to track total pin size is to see if there's unexpected divergence from the content size * replication factor. There's other ways we can do this (sum up what's in our Clusters, do periodic testing to check that new CIDs are in the Cluster x number of times). Don't think this is critical enough to sink too much time into or expose us to the risk we've seen.

But will defer to @alanshaw and @mikeal if there are technical considerations here I'm missing!

**Comment by alanshaw** [comment]
I think we should stop collecting these DAG size metrics. We don't collect them in NFT.Storage. The source of truth for DAG sizes should be dagcargo, where the sizes are calculated and _known_ to be correct. The DAG size stored in our DB is informative (only used in output from the `/status` API) but is not always known immediately (even more so when the pinning service API is shipped) and where we don't know the information up front it comes from dagcargo anyway!

> From a metrics standpoint, I think the main reason we'd want to track total pin size is to see if there's unexpected divergence from the content size * replication factor. There's other ways we can do this (sum up what's in our Clusters, do periodic testing to check that new CIDs are in the Cluster x number of times).

Agree, although the size of the DAG will be the same at each location it is pinned at, so a count of CIDs is probably sufficient.

We _should_ be able to trust that IPFS Cluster is managing pin allocation (that is literally it's job) so really for pinned items I just want to know we have `X` CIDs pinned in our cluster and `Y` CIDs pinned in Pinata (for example). At least that's the metric I'm interested in, if `X != Y`. We don't have this metric in Web3.Storage, but we do in NFT.Storage.

I _think_ the DAG size for non-pinned items is interesting only when debugging so it's not really an everyday metric we want to record. A count of non-pinned items is a useful metric, which we _already collect_ 🙂.

The total DAG size for _pinned_ items, divided by our replication factor (3), will be _approximately_ the DAG size we report for all content in `content` table, minus the DAG size for non-pinned items (which should be minimal).

In summary:

1. We should stop summing up the DAG size per pin.
1. I don't believe we need the alternative because the useful information from that value is going to be approximately the same value as a metric we already collect, assuming our non-pinned items stay low (which they should!).
